src/systemjs-production-loader.js

```diff
@@ -242,31 +242,25 @@ function instantiateIfWasm (url) {
     // detect by leading bytes
     if (bytes[0] === 0 && bytes[1] === 97 && bytes[2] === 115) {
       return WebAssembly.compile(bytes).then(function (m) {
-        /* TODO handle imports when `WebAssembly.Module.imports` is implemented
-        if (WebAssembly.Module.imports) {
-          var deps = [];
-          var setters = [];
-          var importObj = {};
-          WebAssembly.Module.imports(m).forEach(function (i) {
-            var key = i.module;
-            setters.push(function (m) {
-              importObj[key] = m;
-            });
-            if (deps.indexOf(key) === -1)
-              deps.push(key);
+        var deps = [];
+        var setters = [];
+        var importObj = {};
+        WebAssembly.Module.imports(m).forEach(function (i) {
+          var key = i.module;
+          setters.push(function (m) {
+            importObj[key] = m;
           });
-          loader.register(deps, function (_export) {
-            return {
-              setters: setters,
-              execute: function () {
-                _export(new WebAssembly.Instance(m, importObj).exports);
-              }
-            };
-          });
-        }*/
-        // for now we just load WASM without dependencies
-        var wasmModule = new WebAssembly.Instance(m, {});
-        return new ModuleNamespace(wasmModule.exports);
+          if (deps.indexOf(key) === -1)
+            deps.push(key);
+        });
+        loader.register(deps, function (_export) {
+          return {
+            setters: setters,
+            execute: function () {
+              _export(new WebAssembly.Instance(m, importObj).exports);
+            }
+          };
```

Comment:
The way this will work out for circular cases is just providing an undefined value for exports not yet defined by execution. Interestingly a WASM module importing a JS module in a cycle, will actually support function export hoisting though. Whether that happens in actual engines will be yet to be seen.